File: rust/testcode/sqli_014.rs
//! CWE-89: Multi-column INSERT with all bind parameters.

// vuln-code-snippet start testcodeSqli014
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let name = req.param("name");
    let email = req.param("email");
    let role = req.param("role");


    let _query = "INSERT INTO users (name, email, role) VALUES (?, ?, ?)"; // vuln-code-snippet target-line testcodeSqli014
    // In production: sqlx::query(_query).bind(&name).bind(&email).bind(&role).execute(pool).await

    super::shared::BenchmarkResponse::ok(&format!("Inserted user: {}", name))
}
// vuln-code-snippet end testcodeSqli014

File: php/testcode/sqli_014.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_sqli_order_by_whitelist
function sqli014(BenchmarkRequest $req): BenchmarkResponse {
    $pdo = getDbConnection();
    $sortCol = $req->param('sort');
    $allowed = ['id', 'name', 'email', 'created_at'];
    $safeSort = in_array($sortCol, $allowed, true) ? $sortCol : 'id'; // vuln-code-snippet safe-line php_sqli_order_by_whitelist
    $query = "SELECT * FROM users ORDER BY " . $safeSort;
    $result = $pdo->query($query);
    $rows = $result->fetchAll();
    return BenchmarkResponse::json($rows);
}
// vuln-code-snippet end php_sqli_order_by_whitelist
